### PR TITLE
cluster+query: Refactored cluster package and put joins in run.Groups for commands.

### PR DESCRIPTION
**Pull request by Bplotka**
This fixes reloader deadlock.

Depends on https://github.com/improbable-eng/thanos/pull/243

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: pkg/cluster/cluster.go [hunk 1]
```diff
@@ -107,6 +107,8 @@ func New(
 		if err != nil {
 			return nil, errors.Wrap(err, "invalid advertise address, wrong port")
 		}
+	} else if bindHost == "" {
```
[the remark this answers is not in the document]
Comment (fabxc) [reply pkg/cluster/cluster.go]: I can kinda understand what you mean. OTOH this adds complexity that I doubt will come up in deployments soon. That would essentially only help if someone wanted to expose cluster and grpc functionality on different ports, right?

What about the HTTP debugging port then? To be consistent we'd also add a flag to specify another IP address for that one. Looks like this gets easily out of hand and I'd rather avoid it until there's a proven use case.

Maybe leave things for now to get some more user feedback to figure out what we really need?

**Comment by Bplotka** [comment]
PTAL - rebased to master

**Comment by mattbostock** [comment]
Completely agree about being consistent between cluster, HTTP and gRPC
ports.

I think that if we don't want to support that many configuration options to
provide flexible routing, we should maybe remove the
--cluster-advertise-address option for consistency. Doing so would probably
simplify a lot of the code in the cluster package also.

**Comment by fabxc** [comment]
If we propagate our identity via gossip we need a way to specify an IP
there though. Usually that would be provided by an SD system, but we don't
use that in this case. Generally the address gets inferred correctly – but
that may not be true for all setups.

**Comment by mattbostock** [comment]
The IP can be specified using the 'bind' flag. The same is true for the
gRPC and HTTP ports.

I'm personally in favour of adding 'advertise flags' for all ports shared
across the cluster, but if we don't want to support that I think we should
at least be consistent.

**Comment by fabxc** [comment]
Mh, yea – so if we agree that there should be one advertised IP shared
across all ports, that would still be in `--cluster.*` because that's where
we advertise it to. However, we need to advertise the port of the cluster
port regardless as well.

If we just define that the cluster advertise IP is also to be inferred as
the IP for grpc, I don't see a particular consistency problem as it's just
what it's defined as.
We could of course just gossip a grpc port instead of a full grpc address –
that would be a minor implementation detail to me though.

**Comment by fabxc** [comment]
👍

**Comment by mattbostock** [comment]
@fabxc: I disagree with having `--cluster.advertise-address`, `--grpc.address` and `--http.address` then using the host part from `--cluster.advertise-address` as the host advertised for `--grpc.address` and `--http.address` across the cluster, while using the ports from `--grpc.address` and `--http.address`, I think it's too confusing and implicit from a user perspective.

We could change `--cluster.advertise-address` to `--cluster.advertise-host`, which would perhaps be a little clearer (then at least the port is always taken from the `--foo.address` flag for all ports).

My preference is still to have 'advertise' flags for all 3 ports. The three advertise flags would be optional for users that don't need them.

**Comment by Bplotka** [comment]
Ok @mattbostock , @fabxc, can we move this discussion, to a new issue? I might think the problem is not in a scope for this PR.

**Pull request merged**